feat: Feast Security Model (aka RBAC) (#4380)
feast-dev/feast: docs/getting-started/concepts/permission.md

# Permission

## Overview

The Feast permissions model allows to configure granular permission policies to all the resources defined in a feature store.

The configured permissions are stored in the Feast registry and accessible through the CLI and the registry APIs.

The permission authorization enforcement is performed when requests are executed through one of the Feast (Python) servers
- The online feature server (REST)
- The offline feature server (Arrow Flight)
- The registry server (gRPC)

Note that there is no permission enforcement when accessing the Feast API with a local provider.

## Concepts

The permission model is based on the following components:
- A `resource` is a Feast object that we want to secure against unauthorized access.
  - We assume that the resource has a `name` attribute and optional dictionary of associated key-value `tags`.
- An `action` is a logical operation executed on the secured resource, like:
  - `create`: Create an instance.
  - `describe`: Access the instance state.
  - `update`: Update the instance state.
  - `delete`: Delete an instance.

>   - `query`:  Query both online and offline stores.

Reviewer: shouldn't this be `read`?

Reviewer 2: looks like we misunderstood each other. because `read` can be interpreted as both "access state" and "read data", my suggestion was not to use `read` anymore at all. Instead we will have:

- `describe` - this means that a user is allowed to read info from the registry about the feature view.
- `query` - this means that a user is allowed to read dataset from online/offline stores.

cc @dmartinol

Reviewer: got it. I'm okay with that, it just doesn't fit the CRUD convention, right?

Reviewer 2: yeah, it doesn't, but results in less ambiguous naming, which was the goal. plus, describe is also widely used in such cases (in aws iam, at least), so I think that should be fine as well.

  - `query_online`:  Query the online store.
  - `query_offline`:  Query the offline store.

>   - `write`:  Query on any store.

Reviewer 2: typo: write

- `name_pattern`: A regex to match the resource name. Defaults to `None`, meaning that no name filtering is applied
- `required_tags`: Dictionary of key-value pairs that must match the resource tags. Defaults to `None`, meaning that no tags filtering is applied.
- `actions`: The actions authorized by this permission. Defaults to `ALL_VALUES`, an alias defined in the `action` module.
- `policy`: The policy to be applied to validate a client request.

To simplify configuration, several constants are defined to streamline the permissions setup:
- In module `feast.feast_object`:
  - `ALL_RESOURCE_TYPES` is the list of all the `FeastObject` types.
  - `ALL_FEATURE_VIEW_TYPES` is the list of all the feature view types, including those not inheriting from `FeatureView` type like 
  `OnDemandFeatureView`.
- In module `feast.permissions.action`:
  - `ALL_ACTIONS` is the list of all managed actions.
  - `QUERY` includes all the query actions for online and offline store.
  - `WRITE` includes all the write actions for online and offline store.

>   - `CRUD` includes all the state management actions to create, read, update or delete a Feast resource.

Reviewer 2: `describe` instead of `read`

  `OnDemandFeatureView`.
- In module `feast.permissions.action`:
  - `ALL_ACTIONS` is the list of all managed actions.
  - `READ` includes all the read actions for online and offline store.
  - `WRITE` includes all the write actions for online and offline store.
  - `CRUD` includes all the state management actions to create, describe, update or delete a Feast resource.

Given the above definitions, the feature store can be configured with granular control over each resource, enabling partitioned access by 
teams to meet organizational requirements for service and data sharing, and protection of sensitive information.

The `feast` CLI includes a new `permissions` command to list the registered permissions, with options to identify the matching resources for each configured permission and the existing resources that are not covered by any permission.

{% hint style="info" %}
**Note**: Feast resources that do not match any of the configured permissions are not secured by any authorization policy, meaning any user can execute any action on such resources.
{% endhint %}

## Definition examples

> This permission definition grants access to the resource state and the ability to query all the stores for any feature view or

Reviewer: ```suggestion
This permission definition grants access to the resource state and the ability to read all of the stores for any feature view or
```

Reviewer 3: done

> feature service to all users with role `super-reader`:

Reviewer 3: done

> # Permission

Reviewer: ```suggestion
# Role Based Access Control (RBAC) / Permissions
```

## Overview

> The Feast permissions model allows to configure granular permission policies to all the resources defined in a feature store.

Reviewer: ```suggestion
The Feast RBAC model allows for the configuration of granular permission policies to all of the resources defined in a feature store.
```

The configured permissions are stored in the Feast registry and accessible through the CLI and the registry APIs.

> The permission authorization enforcement is performed when requests are executed through one of the Feast (Python) servers

Reviewer: ```suggestion
The RBAC enforcement is performed when requests are executed through one of the Feast (Python) servers
```

- The online feature server (REST)
- The offline feature server (Arrow Flight)

> - The registry server (grpc)

Reviewer: ```suggestion
- The registry server (gRPC)
```

> On the contrary, there is no permission enforcement when accessing the Feast API with a local provider.

Reviewer: ```suggestion
Note that there is no permission enforcement when accessing the Feast API with a local provider.
```

Reviewer 4: Incorporating it.

## Concepts

The permission model is based on the following components:
- A `resource` is a Feast object that we want to secure against unauthorized access.
  - We assume that the resource has a `name` attribute and optional dictionary of associated key-value `tags`.
- An `action` is a logical operation executed on the secured resource, like:
  - `create`: Create an instance.

>   - `read`: Access the instance state.

Reviewer: It is not obvious to me what the difference is between `read` and `query`. What do we mean "access" the instance state? Can you elaborate?

Reviewer 4: I agree it needs to be clarified a bit more.  query is able to query all the feature view meta data(such as name) where as read will allow the user to read actual feature view data from online_store. I will update the PR.

Reviewer: I think we'll need to revisit that naming convention to make it more obvious.

Reviewer: @tokoko what do you think here?

Reviewer 2: This might have been my suggestion :smile: `read` refers to R in CRUD and `query` is the next best name for data retrieval. Maybe we can change `read` to `describe` and we will end up with `describe` and `query`. Should be much less ambiguous.

Author: Yes, @tokoko is definitely the ghost writer behind this design 😉 
We can rename it with low effort but I would keep the utility constant CRUD as is, w/o renaming it to `CDUD`

Reviewer 2: yeah, I'm fine with keeping `CRUD` intact. We can make the change to `describe` if @franciscojavierarceo is also in agreement.

Reviewer: Yeah I do think `describe` is more intuitive. I definitely appreciate the `query` language but I think following `CRUD` is a good enough convention that it'd be better to follow that.

Author: Addressing this changein next commit

  - `update`: Update the instance state.
  - `delete`: Delete an instance.
  - `query`:  Query both online and offline stores.
  - `query_online`:  Query the online store.

>   - `query_offline`:  Query the o ffline store.

Reviewer 4: Incorporating it.

- `name_pattern`: A regex to match the resource name. Defaults to `None`, meaning that no name filtering is applied
- `required_tags`: Dictionary of key-value pairs that must match the resource tags. Defaults to `None`, meaning that no tags filtering is applied.
- `actions`: The actions authorized by this permission. Defaults to `ALL_VALUES`, an alias defined in the `action` module.
- `policy`: The policy to be applied to validate a client request.

Given the above definitions, the feature store can be configured with granular control over each resource, enabling partitioned access by 
teams to meet organizational requirements for service and data sharing, and protection of sensitive information.

The `feast` CLI includes a new `permissions` command to list the registered permissions, with options to identify the matching resources for each configured permission and the existing resources that are not covered by any permission.

{% hint style="info" %}
**Note**: Feast resources that do not match any of the configured permissions are not secured by any authorization policy, meaning any user can execute any action on such resources.
{% endhint %}

## Configuration examples

> This permission configuration allows to access the resource state and query all the stores for any feature view or feature service

Reviewer: ```suggestion
This permission configuration allows access to the resource state and the ability to query all the stores for any feature view or feature service
```

Reviewer 4: I will incorporate it.

to all users with role `super-reader`:
```py

> Permission(

Reviewer: Is this referring to a an instantiation of the FeatureView class? 

I would recommend creating a full [Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example).

    name="feature-reader",

>     types=[FeatureView, FeatureService],

Reviewer: Is this referring to a an instantiation of the FeatureView class? 

I would recommend creating a full [Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example).

Author: ATM we've prepared a reproducible example including a fully distributed environment, in the `examples/rbac` folder, working on a K8s cluster with detailed instructions. 
To support users who are not familiar with K8s, we can prepare another example running all the servers from the local environment and using an OIDC authorization for which we can spin a Keycloak cluster. This is similar to what we did in our POC some time ago.

Would that be enough?

Reviewer: Yeah I think so!

Author: @franciscojavierarceo pls TAL [here](https://github.com/feast-dev/feast/tree/4e778a7f9c755f9b3e0dbae932f199916426a78f/examples/rbac-local), an example driven by notebooks to build a fully operational, distributed system with OIDC authorization.

    policy=RoleBasedPolicy(roles=["super-reader"]),
    actions=[AuthzedAction.READ, QUERY],
)
```
Please note that all sub-classes of `FeatureView` are also included since the default for the `with_subclasses` parameter is `True`.

> This example grants permission to write on all the data sources with `risk_level` tag set to `hi` only to users with role `admin` or `data_team`:

Reviewer: I think you mean "high"

Reviewer 4: Incorporating it.

    actions=[AuthzedAction.WRITE],
)
```

{% hint style="info" %}
**Note**: When using multiple roles in a role-based policy, the user must be granted at least one of the specified roles.
{% endhint %}

The following permission grants authorization to query the offline store of all the feature views including `risky` in the name, to users with role `trusted`:
```py
Permission(
    name="reader",
    types=[FeatureView],
    with_subclasses=False, # exclude sub-classes
    name_pattern=".*risky.*",
    policy=RoleBasedPolicy(roles=["trusted"]),
    actions=[AuthzedAction.QUERY_OFFLINE],
)
```

## Authorizing Feast clients
If you would like to leverage the permissions' functionality then `auth` config block should be added to feature_store.yaml. Currently, feast supports oidc and kubernetes rbac authentication/authorization. 

The default configuration if you don't mention the auth configuration is no_auth configuration.

* No auth configuration example:
```yaml
project: foo
registry: "registry.db"
provider: local
online_store:
    path: foo
entity_key_serialization_version: 2
auth:
  type: no_auth
```

* Kubernetes rbac authentication/authorization example:
{% hint style="info" %}

> **NOTE**: This configuration will only work if you deploy feast on Openshift or Kubernetes platform.

Reviewer: ```suggestion
**NOTE**: This configuration will only work if you deploy feast on Openshift or a Kubernetes platform.
```

Reviewer 4: incorporating it.

to all users with role `super-reader`:
```py
Permission(
    name="feature-reader",
    types=[FeatureView, FeatureService],
    policy=RoleBasedPolicy(roles=["super-reader"]),
    actions=[AuthzedAction.READ, QUERY],
)
```
Please note that all sub-classes of `FeatureView` are also included.

This example grants permission to write on all the data sources with `risk_level` tag set to `high` only to users with role `admin` or `data_team`:
```py
Permission(
    name="ds-writer",
    types=[DataSource],
    required_tags={"risk_level": "high"},
    policy=RoleBasedPolicy(roles=["admin", "data_team"]),
    actions=[AuthzedAction.WRITE],
)
```

{% hint style="info" %}
**Note**: When using multiple roles in a role-based policy, the user must be granted at least one of the specified roles.
{% endhint %}

The following permission grants authorization to query the offline store of all the feature views including `risky` in the name, to users with role `trusted`:
```py
Permission(
    name="reader",
    types=[FeatureView],
    name_pattern=".*risky.*",
    policy=RoleBasedPolicy(roles=["trusted"]),
    actions=[AuthzedAction.QUERY_OFFLINE],
)
```

## Authorizing Feast clients

> If you would like to leverage the permissions' functionality then `auth` config block should be added to feature_store.yaml. Currently, feast supports oidc and kubernetes rbac authentication/authorization. 

Reviewer 2: Maybe also mention that only one mechanism can be configured on the server?

Reviewer 3: updated. Please take another look Thanks
